indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/CostBasedAutoScaler.java

```diff
@@ -0,0 +1,360 @@
+      ServiceEmitter emitter
+  )
+  {
+    this.config = config;
+    this.spec = spec;
+    this.supervisor = supervisor;
+    this.supervisorId = spec.getId();
+    this.emitter = emitter;
+
+    this.costFunction = new WeightedCostFunction();
+
+    this.autoscalerExecutor = Execs.scheduledSingleThreaded("CostBasedAutoScaler-" + StringUtils.encodeForFormat(spec.getId()));
+    this.metricBuilder = ServiceMetricEvent.builder()
+                                           .setDimension(DruidMetrics.SUPERVISOR_ID, supervisorId)
+                                           .setDimension(
+                                               DruidMetrics.STREAM,
+                                               this.supervisor.getIoConfig().getStream()
+                                           );
+  }
+
+  @Override
+  public void start()
+  {
+    Callable<Integer> scaleAction = () -> computeOptimalTaskCount(this.collectMetrics());
+    Runnable onSuccessfulScale = () -> {
+    };
+
+    autoscalerExecutor.scheduleAtFixedRate(
+        supervisor.buildDynamicAllocationTask(scaleAction, onSuccessfulScale, emitter),
+        config.getScaleActionPeriodMillis(),
+        config.getScaleActionPeriodMillis(),
+        TimeUnit.MILLISECONDS
+    );
+
+    log.info(
+        "CostBasedAutoScaler started for supervisorId[%s]: evaluating scaling every [%d]ms",
+        supervisorId,
+        config.getScaleActionPeriodMillis()
+    );
+  }
+
+  @Override
+  public void stop()
+  {
+    autoscalerExecutor.shutdownNow();
+    log.info("CostBasedAutoScaler stopped for supervisorId [%s]", supervisorId);
+  }
+
+  @Override
+  public void reset()
+  {
+    // No-op.
+  }
+
+  private CostMetrics collectMetrics()
+  {
+    if (spec.isSuspended()) {
+      log.debug("Supervisor [%s] is suspended, skipping a metrics collection", supervisorId);
+      return null;
+    }
+
+    final LagStats lagStats = supervisor.computeLagStats();
+    if (lagStats == null) {
+      log.debug("Lag stats unavailable for supervisorId [%s], skipping collection", supervisorId);
+      return null;
+    }
+
+    final int currentTaskCount = supervisor.getIoConfig().getTaskCount();
+    final int partitionCount = supervisor.getPartitionCount();
+
+    final Map<String, Map<String, Object>> taskStats = supervisor.getStats();
+    final double movingAvgRate = extractMovingAverage(taskStats, DropwizardRowIngestionMeters.ONE_MINUTE_NAME);
+    final double pollIdleRatio = extractPollIdleRatio(taskStats);
+
+    final double avgPartitionLag = lagStats.getAvgLag();
+
+    // Use an actual 15-minute moving average processing rate if available
+    final double avgProcessingRate;
+    if (movingAvgRate > 0) {
+      avgProcessingRate = movingAvgRate;
+    } else {
+      // Fallback: estimate processing rate based on idle ratio
+      final double utilizationRatio = Math.max(0.01, 1.0 - pollIdleRatio);
+      avgProcessingRate = config.getDefaultProcessingRate() * utilizationRatio;
```

Comment:
It might be weird to have this be fed from a config (even as a fallback mechanism).

In the future, we can consider computing this based on the stats of previously completed tasks of this supervisor.
OR just use the last known processing rate.

Reply:
Yeah, -1 makes sense too. Best to skip scaling for now rather than make a poor decision.